luarules/gadgets/unit_builder_priority.lua

```diff
@@ -82,101 +104,110 @@ local updateFrame = {}
 local teamList
 local deadTeamList = {}
 local canPassive = {} -- canPassive[unitDefID] = nil / true
-local cost = {} -- cost[unitDefID] = {metal=value,energy=value}
+
+-- Uses a flattened table as per costIDOverride
+local cost = {} -- cost = {unitDefID = metal, (unitDefID + energyOffset) = energy, (unitDefID+buildTimeOffset) = buildtime}, this is now keyed on integers for better cache
+-- Is there any point in the approximate 100Kb of RAM savings? Probably no 
+
 local unitBuildSpeed = {}
 for unitDefID, unitDef in pairs(UnitDefs) do
-    if unitDef.buildSpeed > 0 then
-        unitBuildSpeed[unitDefID] = unitDef.buildSpeed
-    end
-    -- build the list of which unitdef can have low prio mode
-    canPassive[unitDefID] = ((unitDef.canAssist and unitDef.buildSpeed > 0) or #unitDef.buildOptions > 0)
-    cost[unitDefID] = { buildTime = unitDef.buildTime }
-    for _,resName in pairs(resTable) do
-        cost[unitDefID][resName] = unitDef[resName .. "Cost"]
-    end
+	if unitDef.buildSpeed > 0 then
+		unitBuildSpeed[unitDefID] = unitDef.buildSpeed
+	end
+	-- build the list of which unitdef can have low prio mode
+	canPassive[unitDefID] = ((unitDef.canAssist and unitDef.buildSpeed > 0) or #unitDef.buildOptions > 0)
+	cost[unitDefID				  ] = unitDef.metalCost
+	cost[unitDefID +	energyOffset] = unitDef.energyCost
+	cost[unitDefID + buildTimeOffset] = unitDef.buildTime
+	
 end
 
 local function updateTeamList()
 	teamList = spGetTeamList()
 end
 
 function gadget:Initialize()
-    for _,unitID in pairs(Spring.GetAllUnits()) do
-        gadget:UnitCreated(unitID, Spring.GetUnitDefID(unitID), Spring.GetUnitTeam(unitID))
-    end
+	for _,unitID in pairs(Spring.GetAllUnits()) do
+		gadget:UnitCreated(unitID, Spring.GetUnitDefID(unitID), Spring.GetUnitTeam(unitID))
+	end
 	updateTeamList()
 end
 
 function gadget:UnitCreated(unitID, unitDefID, teamID)
-    if canPassive[unitDefID] or unitBuildSpeed[unitDefID] then
-        canBuild[teamID] = canBuild[teamID] or {}
-        canBuild[teamID][unitID] = true
-        realBuildSpeed[unitID] = unitBuildSpeed[unitDefID] or 0
-    end
-    if canPassive[unitDefID] then
-        spInsertUnitCmdDesc(unitID, cmdPassiveDesc)
-        if not passiveCons[teamID] then passiveCons[teamID] = {} end
-        passiveCons[teamID][unitID] = spGetUnitRulesParam(unitID,ruleName) == 1 or nil
-        currentBuildSpeed[unitID] = realBuildSpeed[unitID]
-        spSetUnitBuildSpeed(unitID, currentBuildSpeed[unitID]) -- to handle luarules reloads correctly
-    end
-
-    costID[unitID] = cost[unitDefID]
+	if canPassive[unitDefID] or unitBuildSpeed[unitDefID] then
+		canBuild[teamID] = canBuild[teamID] or {}
+		canBuild[teamID][unitID] = true
+		maxBuildSpeed[unitID] = unitBuildSpeed[unitDefID] or 0
+	end
+	if canPassive[unitDefID] then
+		spInsertUnitCmdDesc(unitID, cmdPassiveDesc)
+		if not passiveCons[teamID] then passiveCons[teamID] = {} end
+		passiveCons[teamID][unitID] = spGetUnitRulesParam(unitID,ruleName) ~= 1 or nil
```

Comment:
This is incorrect. The old code used `spGetUnitRulesParam(unitID,ruleName) == 1 or nil` and this uses `spGetUnitRulesParam(unitID,ruleName) ~= 1 or nil` which seems like a mistake. Though the old code has problems as mentioned in your description, it did correctly handle which builders were high or low priority correctly, while this version doesn't. And changing this equality check resolves seemingly all the problems with priority being incorrectly attributed.

There's more I don't understand here though - this rule is missing for all builders, so everything defaults to being a passive con according to logs place here. It's not clear to me where units get assigned their default priority, e.g. com and all cons are high priority by default, and nanos are low priority by default - it is not defined here, and it does not seem to be defined in the unitdefs either.

Reply:
Ok then this is deeper than i thought, because on a luarules reload, the check is inverted.